src/viam/sdk/common/proto_value.hpp

```diff
@@ -0,0 +1,324 @@
+struct all_proto_moves_noexcept
+    : std::integral_constant<bool,
+                             std::is_nothrow_move_constructible<std::vector<move_may_throw>>{} &&
+                                 std::is_nothrow_move_constructible<
+                                     std::unordered_map<std::string, move_may_throw>>{}> {};
+
+}  // namespace impl
+
+/// @brief Type-erased value for storing google::protobuf::Value types.
+/// A ProtoValue can be nullptr, bool, int, double, std::string, or, recursively, a vector or
+/// string-map of ProtoValue.
+/// This type is used at API/ABI boundaries for interfacing with grpc/proto code.
+class ProtoValue {
+   public:
+    /// @brief Construct a null object.
+    ProtoValue() noexcept;
+
+    /// @brief Construct a nonempty object.
+    template <typename T>
+    ProtoValue(T t) noexcept(std::is_nothrow_move_constructible<T>{});
+
+    /// @brief Deduction helper constructor for string from string literal
+    ProtoValue(const char* str);
+
+    /// @brief Move construct this from other, leaving other in its unspecified-but-valid moved from
+    /// state.
+    ProtoValue(ProtoValue&& other) noexcept(impl::all_proto_moves_noexcept{});
+
+    ProtoValue(const ProtoValue& other);
+
+    /// @brief Move assignment from other, leaving other in its unspecified-but-valid moved from
+    /// state.
+    ProtoValue& operator=(ProtoValue&& other) noexcept(impl::all_proto_moves_noexcept{});
+
+    ProtoValue& operator=(const ProtoValue& other);
+
+    ~ProtoValue();
+
+    /// @brief Test equality of two types.
+    /// @note "Intuitive" arithmetic equality is not supported, but could be.
+    /// Thus, bool{false}, int{0}, and double{0.0} do not compare equal.
+    friend bool operator==(const ProtoValue& lhs, const ProtoValue& rhs);
+
+    void swap(ProtoValue& other) noexcept(impl::all_proto_moves_noexcept{});
+
+    /// @brief Construct from proto value
+    /// @note This method is trivially templated to insulate google::protobuf::Value from our
+    /// API/ABI. It is meant to be called with no template parameters in a translation unit which
+    /// includes <google/protobuf/struct.pb.h>
+    template <typename Value = google::protobuf::Value>
+    static ProtoValue from_proto(const Value& v);  // NOLINT(misc-no-recursion)
+
+    friend void to_proto(const ProtoValue& t, google::protobuf::Value* v);
+
+    /// @name Casting API
+    ///@{
+
+    /// @brief Obtain integer constant representing the stored data type.
+    int kind() const;
+
+    /// @brief Checks whether this ProtoT is an instance of type T.
+    template <typename T>
+    bool is_a() const;
+
+    /// @brief Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T* dyn_cast(ProtoValue&);
+
+    /// @brief Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T const* dyn_cast(const ProtoValue&);
```

Comment:
Are these, effectively, the main accessors? Why make them friend casts rather than just regular `get` members?

Reply:
They are, yes! It's kind of semantic and I'm not opposed to calling it `get` instead, I suppose `get` is a better fit for the mental model of "this is kind of like a JSON value type" rather than casting which evokes polymorphism.